File: app/src/main/java/com/vtop/ui/LoginActivity.java
package com.vtop.ui;

import android.content.Intent;
import android.os.Bundle;
import android.text.InputType;
import android.util.Log;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.vtop.network.VtopClient;
import com.vtop.utils.Vault;
import com.vtop.models.*;

import java.util.List;
import java.util.Map;

@SuppressWarnings("SpellCheckingInspection")
public class LoginActivity extends AppCompatActivity {

    private VtopClient activeClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (Vault.getTimetable(this) != null) {
            startActivity(new Intent(this, MainActivity.class));
            finish();
            return;
        }

        FrameLayout root = new FrameLayout(this);
        root.setTag("vtop_root");
        root.setLayoutParams(new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
        setContentView(root);

        String[] savedCreds = Vault.getCredentials(this);

        com.vtop.ui.VtopLoginBridge.launchAuth(root, savedCreds[0], savedCreds[1], new com.vtop.ui.AuthActionCallback() {
            @Override
            public void onLoginSubmit(@org.jetbrains.annotations.NotNull String regNo, @org.jetbrains.annotations.NotNull String pass) {
                executeLoginSequence(regNo, pass);
            }

            @Override
            public void onSemesterSelect(@org.jetbrains.annotations.NotNull String semId, @org.jetbrains.annotations.NotNull String semName) {
                Vault.saveSelectedSemester(LoginActivity.this, semId, semName);
                executeDataDownload(semId);
            }
        });
    }

    private void executeLoginSequence(String reg, String pwd) {
        com.vtop.ui.VtopLoginBridge.INSTANCE.getCurrentState().setValue(com.vtop.ui.AuthState.LOADING_SEMESTERS);
        com.vtop.ui.VtopLoginBridge.INSTANCE.getLoginError().setValue("Login succeeded but couldn't load semesters. Please try again.");

        new Thread(() -> {
            activeClient = new VtopClient(LoginActivity.this, reg, pwd);
            boolean isSuccess = false;
            int attempts = 0;
            String finalError = "Network timeout or portal down. Please try again.";

            while (attempts < 3 && !isSuccess) {
                attempts++;
                try {
                    isSuccess = activeClient.autoLogin(this, new VtopClient.LoginListener() {
                        @Override
                        public void onStatusUpdate(String message) {
                            Log.d("VTOP_LOGIN_STATUS", message);
                        }

                        @Override
                        public void onOtpRequired(VtopClient.OtpResolver resolver) {
                            Log.d("VTOP_OTP", "onOtpRequired called — showing OTP form");
                            runOnUiThread(() -> {
                                Log.d("VTOP_OTP", "Adding OTP overlay to root");
                                FrameLayout root = findViewById(android.R.id.content);
                                com.vtop.ui.VtopTimetableComposeUIKt.showOtpOverlay(root, LoginActivity.this, resolver);
                            });
                        }
                    });
                } catch (Exception e) {
                    Log.e("VTOP_LOGIN", "Attempt " + attempts + " failed", e);
                    if (e.getMessage() != null) {
                        finalError = e.getMessage().replace("java.lang.Exception:", "").trim();
                    } else {
                        finalError = "Unknown Authentication Error";
                    }

                    // Fixed: Checking for "Canceled" with one 'L'
                    if (finalError.contains("Login/Password") || finalError.contains("Invalid Login") || finalError.contains("Canceled")) {
                        break;
                    }
                }

                if (!isSuccess && attempts < 3 && !finalError.contains("Login/Password") && !finalError.contains("Canceled")) {
                    try { Thread.sleep(2000); } catch (InterruptedException ignored) {}
                }
            }

            if (isSuccess) {
                Vault.saveCredentials(this, reg, pwd);
                List<Map<String, String>> semesters = activeClient.fetchSemesters();

                android.content.SharedPreferences prefs = getSharedPreferences("VTOP_PREFS", android.content.Context.MODE_PRIVATE);
                org.json.JSONArray jsonArray = new org.json.JSONArray();
                for (Map<String, String> sem : semesters) {
                    org.json.JSONObject obj = new org.json.JSONObject();
                    try {
                        obj.put("id", sem.get("id"));
                        obj.put("name", sem.get("name"));
                        jsonArray.put(obj);
                    } catch (org.json.JSONException ignored) {}
                }
                prefs.edit().putString("SEMESTERS_CACHE", jsonArray.toString()).apply();

                runOnUiThread(() -> {
                    com.vtop.ui.VtopLoginBridge.INSTANCE.getFetchedSemesters().setValue(semesters);
                    com.vtop.ui.VtopLoginBridge.INSTANCE.getCurrentState().setValue(com.vtop.ui.AuthState.SELECT_SEMESTER);
                });
            } else {
                final String displayError = finalError;
                runOnUiThread(() -> {
                    com.vtop.ui.VtopLoginBridge.INSTANCE.getCurrentState().setValue(com.vtop.ui.AuthState.FORM);
                    com.vtop.ui.VtopLoginBridge.INSTANCE.getLoginError().setValue(displayError);
                });
            }
        }).start();
    }

    private void executeDataDownload(String semId) {
        com.vtop.ui.VtopLoginBridge.INSTANCE.getCurrentState().setValue(com.vtop.ui.AuthState.DOWNLOADING_DATA);

        new Thread(() -> {
            boolean criticalFailure = false;
            try {
                try {
                    String ttHtml = activeClient.fetchTimetableRawHtml(semId, msg -> {});
                    if (ttHtml != null) {
                        Vault.saveTimetable(this, com.vtop.logic.TimetableParser.parse(ttHtml));
                    } else criticalFailure = true;
                } catch (Exception e) {
                    Log.e("VTOP_SYNC", "Timetable failed", e);
                    criticalFailure = true;
                }

                try {
                    String attHtml = activeClient.fetchAttendanceRawHtml(semId, msg -> {});
                    if (attHtml != null) {
                        List<AttendanceModel> attSummary = com.vtop.logic.AttendanceParser.parseSummary(attHtml);
                        String regNo = Vault.getCredentials(LoginActivity.this)[0];

                        for (AttendanceModel course : attSummary) {
                            try {
                                String detailHtml = activeClient.fetchAttendanceDetailRawHtml(
                                        semId,
                                        course.classId,
                                        course.courseType,
                                        regNo,
                                        msg -> {}
                                );
                                if (detailHtml != null) {
                                    com.vtop.logic.AttendanceParser.parseDetailAndUpdate(detailHtml, course);
                                }
                            } catch (Exception detailEx) {
                                Log.e("VTOP_SYNC", "Failed to fetch detailed history for: " + course.courseCode, detailEx);
                            }
                        }
                        Vault.saveAttendance(LoginActivity.this, attSummary);
                    } else criticalFailure = true;
                } catch (Exception e) {
                    Log.e("VTOP_SYNC", "Attendance global fetch failed", e);
                    criticalFailure = true;
                }

                try {
                    String examsHtml = activeClient.fetchExamScheduleRawHtml(semId, msg -> {});
                    if (examsHtml != null) {
                        Vault.saveExamSchedule(this, com.vtop.logic.ExamScheduleParser.INSTANCE.parse(examsHtml));
                    }
                } catch (Exception e) { Log.e("VTOP_SYNC", "Exams failed", e); }

                try {
                    String marksHtml = activeClient.fetchMarksRawHtml(semId, msg -> {});
                    if (marksHtml != null) {
                        Vault.saveMarks(this, com.vtop.logic.MarksParser.INSTANCE.parseMarks(marksHtml));
                    }
                } catch (Exception e) { Log.e("VTOP_SYNC", "Marks failed", e); }

                try {
                    String gradesHtml = activeClient.fetchGradesRawHtml(semId, msg -> {});
                    if (gradesHtml != null) {
                        Vault.saveGrades(this, com.vtop.logic.MarksParser.INSTANCE.parseGrades(gradesHtml));
                    }
                } catch (Exception e) { Log.e("VTOP_SYNC", "Grades failed", e); }

                try {
                    String historyHtml = activeClient.fetchHistoryRawHtml(msg -> {});
                    if (historyHtml != null) {
                        kotlin.Pair<CGPASummary, java.util.List<GradeHistoryItem>> historyData = com.vtop.logic.MarksParser.INSTANCE.parseHistory(historyHtml);

                        if (historyData.getFirst() != null) Vault.saveCGPASummary(this, historyData.getFirst());
                        if (historyData.getSecond() != null) Vault.saveHistory(this, historyData.getSecond());
                    }
                } catch (Exception e) { Log.e("VTOP_SYNC", "History failed", e); }

                if (criticalFailure) {
                    throw new Exception("Core academic data failed to fetch. Portal may be down.");
                }

                Vault.saveLastSyncTime(this);

                runOnUiThread(() -> {
                    startActivity(new Intent(LoginActivity.this, MainActivity.class));
                    finish();
                });
            } catch (Exception e) {
                Log.e("VTOP_SYNC", "Critical Sync Failure", e);
                final String safeError = e.getMessage() != null ? e.getMessage() : "Unknown Critical Error";
                runOnUiThread(() -> {
                    com.vtop.ui.VtopLoginBridge.INSTANCE.getCurrentState().setValue(com.vtop.ui.AuthState.FORM);
                    Toast.makeText(this, "Sync Error: " + safeError, Toast.LENGTH_LONG).show();
                });
            }
        }).start();
    }
}

File: app/src/main/java/com/vtop/ui/LoginActivity.kt
@file:Suppress("SpellCheckingInspection", "UNUSED_VARIABLE")

package com.vtop.ui

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.runtime.*
import androidx.core.content.edit
import androidx.core.view.WindowCompat
import androidx.lifecycle.lifecycleScope
import com.vtop.network.VtopClient
import com.vtop.core.AppState
import com.vtop.core.AuthStateManager
import com.vtop.core.SessionManager
import com.vtop.sync.SyncManager
import com.vtop.ui.screens.auth.LoginScreen
import com.vtop.ui.screens.auth.OnboardingFlow
import com.vtop.ui.theme.AppTheme
import com.vtop.ui.theme.AppThemeMode
import com.vtop.ui.theme.AuthActionCallback
import com.vtop.ui.theme.AuthState
import com.vtop.utils.Vault
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import kotlin.time.Duration.Companion.seconds

class LoginActivity : ComponentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val sharedPrefs = getSharedPreferences("VTOP_PREFS", MODE_PRIVATE)
        val isExplicitlyLoggedOut = sharedPrefs.getBoolean("IS_EXPLICITLY_LOGGED_OUT", false)

        val credentials = Vault.getCredentials(this)
        val savedReg = credentials[0]
        val savedPass = credentials[1]

        if (!isExplicitlyLoggedOut &&
            !savedReg.isNullOrEmpty() &&
            !savedPass.isNullOrEmpty()
        ) {
            startActivity(Intent(this, MainActivity::class.java))
            finish()
            return
        }

        WindowCompat.setDecorFitsSystemWindows(window, false)

        val savedThemeString = sharedPrefs.getString("APP_THEME", AppThemeMode.SYSTEM.name) ?: AppThemeMode.SYSTEM.name
        val savedTheme = AppThemeMode.valueOf(savedThemeString)

        AuthStateManager.currentState.value = AuthState.FORM

        setContent {
            var showOnboarding by remember { mutableStateOf(false) }

            AppTheme(themeMode = savedTheme) {
                if (showOnboarding) {
                    OnboardingFlow(onComplete = {
                        val intent = Intent(this@LoginActivity, MainActivity::class.java)
                        startActivity(intent)
                        finish()
                    })
                } else {
                    LoginScreen(
                        savedReg = savedReg,
                        savedPass = savedPass,
                        callback = object : AuthActionCallback {

                            override fun onLoginSubmit(regNo: String, pass: String) {
                                Vault.saveCredentials(this@LoginActivity, regNo, pass)
                                sharedPrefs.edit { putBoolean("IS_EXPLICITLY_LOGGED_OUT", false) }

                                AuthStateManager.currentState.value = AuthState.LOADING_SEMESTERS

                                lifecycleScope.launch(Dispatchers.IO) {
                                    try {
                                        val client = VtopClient(this@LoginActivity, regNo, pass)
                                        client.reinitializeSession(this@LoginActivity)

                                        var loginSuccess = false
                                        var attempts = 0
                                        val maxAttempts = 3

                                        while (!loginSuccess && attempts < maxAttempts) {
                                            attempts++
                                            if (attempts > 1) {
                                                withContext(Dispatchers.Main) { AuthStateManager.loginError.value = "Retrying login ($attempts/$maxAttempts)..." }
                                                delay(1.seconds)
                                            }

                                            loginSuccess = client.autoLogin(this@LoginActivity, object : VtopClient.LoginListener {
                                                override fun onStatusUpdate(message: String) { Log.d("VTOP_LOGIN", message) }
                                                override fun onOtpRequired(resolver: VtopClient.OtpResolver) {
                                                    lifecycleScope.launch(Dispatchers.Main) { AppState.currentOtpResolver.value = resolver }
                                                }
                                            })
                                        }

                                        if (loginSuccess) {
                                            // Inject the authorized client into the App Session
                                            SessionManager.setSyncClient(client)

                                            if (client.authorizedId != null && !client.authorizedId.isEmpty() && !client.authorizedId.equals(regNo)) {
                                                Vault.saveRegNo(this@LoginActivity, client.authorizedId)
                                            }

                                            val semestersList = client.fetchSemesters()
                                            val firstSemName = semestersList.firstOrNull()?.get("name") ?: ""
                                            val currentIndex = getActiveSemesterIndex(this@LoginActivity, firstSemName)

                                            val processedSemesters = semestersList.mapIndexed { index, map ->
                                                map.toMutableMap().apply { this["isCurrent"] = (index == currentIndex).toString() }
                                            }

                                            withContext(Dispatchers.Main) {
                                                AuthStateManager.loginError.value = null
                                                AuthStateManager.fetchedSemesters.value = processedSemesters
                                                AuthStateManager.currentState.value = AuthState.SELECT_SEMESTER
                                            }
                                        } else {
                                            withContext(Dispatchers.Main) {
                                                AuthStateManager.loginError.value = "Login failed. VTOP might be unresponsive."
                                                AuthStateManager.currentState.value = AuthState.FORM
                                            }
                                        }
                                    } catch (e: com.vtop.network.VtopException) {
                                        withContext(Dispatchers.Main) {
                                            AuthStateManager.loginError.value = e.message
                                            AuthStateManager.currentState.value = AuthState.FORM
                                        }
                                    } catch (e: Exception) {
                                        withContext(Dispatchers.Main) {
                                            AuthStateManager.loginError.value = "Network error: ${e.message}"
                                            AuthStateManager.currentState.value = AuthState.FORM
                                        }
                                    }
                                }
                            }

                            override fun onSemesterSelect(semId: String, semName: String) {
                                Vault.saveSelectedSemester(this@LoginActivity, semId, semName)

                                // Trigger sync instantly without blocking the UI
                                lifecycleScope.launch(Dispatchers.IO) {
                                    SyncManager.performSync(
                                        context = this@LoginActivity,
                                        priorityTab = "HOME",
                                        forceNewSession = false,
                                        targetSemId = semId,
                                        skipLogin = true // Use the cookies we just established!
                                    )
                                }

                                showOnboarding = true
                            }
                        }
                    )
                }
            }
        }
    }

    private fun getActiveSemesterIndex(context: Context, topSemesterName: String): Int {
        if (topSemesterName.isEmpty()) return 0
        try {
            val jsonString = context.assets.open("academic_calendar.json").bufferedReader().use { it.readText() }
            val jsonObject = JSONObject(jsonString)
            if (!jsonObject.has("blocked_dates")) return 0

            val blocked = jsonObject.getJSONObject("blocked_dates")
            val keys = blocked.keys()

            val sdf = SimpleDateFormat("dd-MM-yyyy", Locale.ENGLISH)
            val today = Calendar.getInstance()
            val currentYear = today.get(Calendar.YEAR)
            val todayDate = sdf.parse(sdf.format(today.time)) ?: return 0

            while (keys.hasNext()) {
                val dateKey = keys.next()
                val eventName = blocked.getString(dateKey)
                val normalizedEvent = eventName.replace("Commencement of ", "", ignoreCase = true).trim()

                if (topSemesterName.contains(normalizedEvent, ignoreCase = true) || eventName.contains(topSemesterName, ignoreCase = true)) {
                    val targetDate = sdf.parse("$dateKey-$currentYear")
                    if (targetDate != null && targetDate.after(todayDate)) {
                        return 1
                    }
                }
            }
        } catch (_: Exception) { }
        return 0
    }
}